File: android-client/lib/services/waiting_room_service.dart
import 'dart:async';

import 'package:android_client/classes/game_room.dart';
import 'package:android_client/classes/user_game.dart';
import 'package:android_client/screens/game_screen.dart';
import 'package:android_client/services/game_service.dart';
import 'package:android_client/services/socket_service.dart';
import 'package:flutter/material.dart';

class WaitingRoomService {
  final SocketService _socketService = SocketService();
  final GameService gameService = GameService();
  GameRoom? gameRoom;
  late String username;
  late String gameMode;
  late String roomId;
  late BuildContext buildContext;
  StreamController<bool> gameCanceled$ = StreamController<bool>();
  StreamController<bool> rejected$ = StreamController<bool>();
  StreamController<bool> accepted$ = StreamController<bool>();
  StreamController<GameRoom?> gameRoomUpdated$ = StreamController<GameRoom?>();
  StreamController<GameRoom> gameRoomApp$ = StreamController<GameRoom>();

  static final WaitingRoomService _instance = WaitingRoomService._internal();

  factory WaitingRoomService() {
    return _instance;
  }

  WaitingRoomService._internal();
  Stream<GameRoom> get getGameRoomStream => gameRoomApp$.stream;

  Stream<bool> get getGameCanceledStream => gameCanceled$.stream;

  Stream<bool> get getRejectedStream => rejected$.stream;

  Stream<bool> get getAcceptedStream => accepted$.stream;

  Stream<GameRoom?> get getGameRoomUpdatedStream => gameRoomUpdated$.stream;

  void reloadGames() {
    gameService.gameMode = gameMode;
  }

  void acceptPlayer(String player, bool isAndroid) {
    _socketService.send('acceptPlayer', <String, Object?>{
      "roomId": gameRoom?.roomId,
      "username": player,
      "isAndroid": isAndroid
    });
  }

  void rejectPlayer(String player) {
    _socketService.send('rejectPlayer',
        <String, String?>{"roomId": gameRoom?.roomId, "username": player});
  }

  void abortGame() {
    if (gameRoom?.userGame.creator == username) {
      _socketService.send('abortGameCreation', gameRoom?.roomId);
    } else if (gameRoom != null) {
      _socketService.send('leaveGame',
          <String, String?>{"roomId": gameRoom?.roomId, "username": username});
    }
  }

  void startGame() {
    gameRoom?.userGame.timer = gameRoom?.gameConstants!.gameDuration ?? 0;

    if (gameRoom?.userGame.creator == username) {
      _socketService.send('startingGame', gameRoom?.toJson());
    }
  }

  void createGame(GameRoom gameRoom) {
    this.gameRoom = gameRoom;
    username = gameRoom.userGame.creator;
    gameMode = gameRoom.gameMode;
    roomId = gameRoom.roomId;
    handleWaitingRoomSocket();
    _socketService.send('createGame', gameRoom.toJson());
    gameRoomUpdated$.add(gameRoom);
  }

  void joinGame(
      String username, String gameMode, String roomId, String? gameName) {
    gameRoom = null;
    this.username = username;
    this.gameMode = gameMode;
    this.roomId = roomId;
    final CurrentPlayer user =
        CurrentPlayer(username: username, isAndroid: true);
    handleWaitingRoomSocket();
    _socketService.send('askingToJoinGame', <String, Object?>{
      "gameName": gameName,
      "user": user,
      "gameMode": gameMode,
      "roomId": roomId
    });
  }

  void observeGame(
      String username, String gameMode, String roomId, String? gameName) {
    gameRoom = null;
    this.username = username;
    this.gameMode = gameMode;
    this.roomId = roomId;
    final CurrentPlayer user =
        CurrentPlayer(username: username, isAndroid: true);
    handleWaitingRoomSocket();
    _socketService.send('askingToObserveGame', <String, Object?>{
      "gameName": gameName,
      "user": user,
      "gameMode": gameMode,
      "roomId": roomId
    });
  }

  void handleWaitingRoomSocket() {
    _socketService.on('startingGame', (data) {
      if (data == null) {
        print(
            'Nous avons eu un problème pour obtenir les informations de jeu du serveur');
        return;
      }
      GameRoom? gameRoomSocket = GameRoom.fromJson(data);
      if (gameRoomSocket.userGame.currentPlayers
          .where((CurrentPlayer user) => user.username == username)
          .any((CurrentPlayer user) => true)) {
        gameRoom = gameRoomSocket;
        accepted$.add(true);
        gameRoom?.started = true;
        gameService.startGame(gameRoom!, username);
        removeSocketListeners();
        _socketService.off('startingGame');
        Navigator.pop(buildContext);
        Navigator.push(
          buildContext,
          MaterialPageRoute(
              builder: (BuildContext buildContext) => const GameScreen()),
        );
      } else {
        rejected$.add(true);
      }
      _socketService.off('startingGame');
    });
    _socketService.on('gameInfo', (data) {
      if (data == null) {
        print(
            'Nous avons eu un problème pour obtenir les informations de jeu du serveur');
        _socketService.off('startingGame');
        return;
      }
      GameRoom? gameRoomSocket = GameRoom.fromJson(data);
      if ((gameRoom == null ||
              gameRoom?.userGame.gameName ==
                  gameRoomSocket.userGame.gameName) &&
          gameMode == gameRoomSocket.gameMode) {
        gameRoom = gameRoomSocket;
        gameRoomUpdated$.add(gameRoom);
        if (gameRoom!.userGame.observers!
            .where((Observer user) => user.username == username)
            .any((Observer user) => true)) {
          gameService.observeGame(gameRoom!, username);
          removeSocketListeners();
          _socketService.off('startingGame');
          Navigator.pop(buildContext);
          Navigator.push(
            buildContext,
            MaterialPageRoute(
                builder: (BuildContext buildContext) => const GameScreen()),
          );
        }
      }
    });
    _socketService.on('playerAccepted', (data) {
      if (data == null) {
        print(
            'Nous avons eu un problème pour obtenir les informations de jeu du serveur');
        return;
      }
      GameRoom? gameRoomSocket = GameRoom.fromJson(data);
      gameRoom = gameRoomSocket;
      gameRoomUpdated$.add(gameRoom);
    });

    _socketService.on('playerRejected', (data) {
      if (data == null) {
        print(
            'Nous avons eu un problème pour obtenir les informations de jeu du serveur');
        return;
      }
      GameRoom? gameRoomSocket = GameRoom.fromJson(data);
      if (!gameRoomSocket.userGame.currentPlayers
              .where((CurrentPlayer user) => user.username == username)
              .any((CurrentPlayer user) => true) &&
          !gameRoomSocket.userGame.potentialPlayers!
              .where((CurrentPlayer user) => user.username == username)
              .any((CurrentPlayer user) => true)) {
        rejected$.add(true);
      } else {
        gameRoom = gameRoomSocket;
        gameRoomUpdated$.add(gameRoom);
      }
    });

    _socketService.on('gameCanceled', (data) {
      if (data == null) {
        print(
            'Nous avons eu un problème pour obtenir les informations de jeu du serveur');
        return;
      }
      GameRoom? gameRoomSocket = GameRoom.fromJson(data);
      if (gameRoom?.userGame.gameName == gameRoomSocket.userGame.gameName &&
          gameRoom!.gameMode == gameMode &&
          (gameRoomSocket.userGame.currentPlayers
                  .where((CurrentPlayer user) => user.username == username)
                  .any((CurrentPlayer user) => true) ||
              gameRoomSocket.userGame.potentialPlayers!
                  .where((CurrentPlayer user) => user.username == username)
                  .any((CurrentPlayer user) => true))) {
        gameCanceled$.add(true);
      }
    });

    _socketService.on('gameDeleted', (data) {
      if (data == null) {
        print(
            'Nous avons eu un problème pour obtenir les informations de jeu du serveur');
        return;
      }
      if (gameMode == 'mode Temps Limité') {
        gameService.slides.removeWhere((slide) => slide.name == data);
        if (gameService.slides.length == 0) {
          gameCanceled$.add(true);
        }
      } else if (gameRoom?.userGame.gameName == data) {
        gameCanceled$.add(true);
      }
    });
  }

  void closeStreams() {
    gameCanceled$.close();
    rejected$.close();
    accepted$.close();
    gameRoomApp$.close();
    gameCanceled$ = StreamController<bool>();
    rejected$ = StreamController<bool>();
    accepted$ = StreamController<bool>();
    gameRoomUpdated$ = StreamController<GameRoom?>();
    gameRoomApp$ = StreamController<GameRoom>();
  }

  void resetGameService() {
    _socketService.off('startingGame');
    gameService.closeStreams();
  }

  void removeSocketListeners() {
    _socketService.off('gameInfo');
    _socketService.off('gameCreated');
    _socketService.off('playerAccepted');
    _socketService.off('playerRejected');
    _socketService.off('gameCanceled');
    _socketService.off('gameDeleted');
  }
}

File: windows-client/src/app/services/waiting-room/waiting-room.service.ts
import { Injectable } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { Router } from '@angular/router';
import { MessageDialogComponent } from '@app/components/message-dialog/message-dialog.component';
import { CommunicationSocketService } from '@app/services/communication-socket/communication-socket.service';
import { GameService } from '@app/services/game/game.service';
import { GameRoom } from '@common/classes/game-room';
import { WaitingRoomEvents } from '@common/enums/waiting-room.gateway.variables';
import { GameMode } from '@common/game-mode';
import { Subject } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class WaitingRoomService {
    rejected$ = new Subject<boolean>();
    accepted$ = new Subject<boolean>();
    gameCanceled$ = new Subject<boolean>();
    gameRoom: GameRoom;
    roomId: string;
    username: string;
    gameMode: string;

    // eslint-disable-next-line max-params
    constructor(
        private router: Router,
        private readonly socketService: CommunicationSocketService,
        private gameService: GameService,
        public dialog: MatDialog,
    ) {}

    reloadGames() {
        this.gameService.gameMode = this.gameMode;
    }

    playerRejected(player: string): void {
        if (this.socketService.isSocketAlive()) {
            this.socketService.send(WaitingRoomEvents.RejectPlayer, { roomId: this.gameRoom.roomId, username: player });
        }
    }

    playerAccepted(player: string, isAndroid: boolean): void {
        if (this.socketService.isSocketAlive()) {
            this.socketService.send(WaitingRoomEvents.AcceptPlayer, { roomId: this.gameRoom.roomId, username: player, isAndroid });
        }
    }

    abortGame(): void {
        if (this.socketService.isSocketAlive() && this.gameRoom?.userGame.creator === this.username) {
            this.socketService.send(WaitingRoomEvents.AbortGameCreation, this.gameRoom.roomId);
        } else if (this.socketService.isSocketAlive() && this.gameRoom) {
            this.socketService.send(WaitingRoomEvents.LeaveGame, { roomId: this.gameRoom.roomId, username: this.username });
        }
        this.router.navigateByUrl('/', { skipLocationChange: true }).then(() => {
            this.router.navigate([this.gameMode === GameMode.classicMode ? '/selection' : '/home']);
        });
    }

    startGame(): void {
        this.gameRoom.userGame.timer = this.gameRoom.gameConstants.gameDuration;
        if (this.gameRoom.userGame.creator === this.username) {
            this.socketService.send(WaitingRoomEvents.StartingGame, this.gameRoom);
        }
    }

    createGame(gameRoom: GameRoom): void {
        this.gameRoom = gameRoom;
        this.username = this.gameRoom.userGame.creator;
        this.gameMode = this.gameRoom.gameMode;
        this.roomId = gameRoom.roomId;
        this.handleWaitingRoomSocket();
        this.socketService.send(WaitingRoomEvents.CreateGame, this.gameRoom);
    }

    // TODO: refactor this
    // eslint-disable-next-line max-params
    joinGame(username: string, gameMode: string, roomId: string, gameName = undefined as unknown as string): void {
        this.gameRoom = undefined as unknown as GameRoom;
        this.username = username;
        this.gameMode = gameMode;
        this.roomId = roomId;
        const user = {
            username,
            isAndroid: false,
        };
        this.handleWaitingRoomSocket();
        this.socketService.send(WaitingRoomEvents.AskingToJoinGame, { gameName, user, gameMode, roomId });
    }

    observeGame(username: string, gameMode: string, roomId: string, gameName = undefined as unknown as string): void {
        this.gameRoom = undefined as unknown as GameRoom;
        this.username = username;
        this.gameMode = gameMode;
        this.roomId = roomId;
        const user = {
            username,
            isAndroid: false,
        };
        this.handleWaitingRoomSocket();
        this.socketService.send(WaitingRoomEvents.AskingToObserveGame, { gameName, user, gameMode, roomId });
    }

    handleWaitingRoomSocket(): void {
        this.socketService.on(WaitingRoomEvents.StartingGame, (gameRoom: GameRoom) => {
            if (gameRoom && gameRoom.userGame.currentPlayers.some((player) => player.username === this.username)) {
                this.gameRoom = gameRoom;
                this.accepted$.next(true);
                this.gameRoom.started = true;
                this.gameService.startGame(this.gameRoom, this.username);
                this.removeSocketListeners();
                this.socketService.off(WaitingRoomEvents.StartingGame);
                this.router.navigate(['/game']);
            } else if (gameRoom) {
                this.rejected$.next(true);
            }
            this.socketService.off(WaitingRoomEvents.StartingGame);
        });

        this.socketService.on(WaitingRoomEvents.GameInfo, (gameRoom: GameRoom) => {
            if (
                gameRoom &&
                (!this.gameRoom || this.gameRoom.userGame.gameName === gameRoom.userGame.gameName) &&
                this.gameMode === gameRoom.gameMode
            ) {
                this.gameRoom = gameRoom;
                if (this.gameRoom.userGame.observers?.some((observer) => observer.username === this.username)) {
                    this.gameService.observeGame(this.gameRoom, this.username);
                    this.removeSocketListeners();
                    this.socketService.off(WaitingRoomEvents.StartingGame);
                    this.router.navigate(['/game']);
                }
            } else if (!gameRoom) {
                this.dialog.open(MessageDialogComponent, {
                    panelClass: 'custom-modal',
                    data: { message: 'Nous avons eu un problème pour obtenir les informations de jeu du serveur' },
                });
            }
        });

        this.socketService.on(WaitingRoomEvents.PlayerAccepted, (gameRoom: GameRoom) => {
            this.gameRoom = gameRoom;
        });

        this.socketService.on(WaitingRoomEvents.PlayerRejected, (gameRoom: GameRoom) => {
            if (
                gameRoom &&
                !gameRoom.userGame.currentPlayers.some((player) => player.username === this.username) &&
                !gameRoom.userGame.potentialPlayers?.some((player) => player.username === this.username)
            ) {
                this.rejected$.next(true);
            } else if (gameRoom) {
                this.gameRoom = gameRoom;
            }
        });

        this.socketService.on(WaitingRoomEvents.GameCanceled, (gameRoom: GameRoom) => {
            if (
                this.gameRoom?.userGame.gameName === gameRoom?.userGame.gameName &&
                this.gameRoom.gameMode === this.gameMode &&
                (gameRoom.userGame.currentPlayers.some((player) => player.username === this.username) ||
                    gameRoom.userGame.potentialPlayers?.some((player) => player.username === this.username))
            ) {
                this.gameCanceled$.next(true);
            }
        });

        this.socketService.on(WaitingRoomEvents.GameDeleted, (gameName: string) => {
            if (this.gameMode === GameMode.limitedTimeMode) {
                this.gameService.slides = this.gameService.slides.filter((slide) => slide.name !== gameName);
                if (this.gameService.slides.length === 0) {
                    this.gameCanceled$.next(true);
                }
            } else if (this.gameRoom.userGame.gameName === gameName) {
                this.gameCanceled$.next(true);
            }
        });
    }

    removeSocketListeners(): void {
        this.socketService.off(WaitingRoomEvents.GameInfo);
        this.socketService.off(WaitingRoomEvents.GameCreated);
        this.socketService.off(WaitingRoomEvents.PlayerAccepted);
        this.socketService.off(WaitingRoomEvents.PlayerRejected);
        this.socketService.off(WaitingRoomEvents.GameCanceled);
        this.socketService.off(WaitingRoomEvents.GameDeleted);
    }
}
